Helidon prematurely terminating WebSocket session when client is JDK HttpClient

## Environment Details
* Helidon Version: 2.0.0+ (Affects all versions)
* Helidon SE 
* JDK version: 11+
* OS: MacOS
* Docker version (if applicable): N/A

----------

## Problem Description

- JDK HttpClient sends [`Content-Length: 0` with GET requests](https://github.com/openjdk/jdk/blob/master/src/java.net.http/share/classes/jdk/internal/net/http/Http1Request.java#L301). Not prohibited by HTTP spec, but definitely unusual.
- This includes the GET request which initiates the WebSocket Upgrade. In the case of the upgrade request, calling code has no way to configure/modify the GET request to prevent the `Content-Length: 0` being added.
- io.helidon.webserver.ForwardingHandler [assumes any request](https://github.com/oracle/helidon/blob/master/webserver/webserver/src/main/java/io/helidon/webserver/ForwardingHandler.java#L370) (regardless of HTTP method) with a `Content-Length: 0` is completed. 

## Expected Behaviour

JDK WebSocket Client able to connect to Helidon WebSocket server and transmit and receive messages in both directions.

## Actual Behaviour

JDK Websocket Client adds spurious Content-Length: 0 header, causing Helidon to prematurely terminate the WebSocket session. Unable to send or receive messages from JDK WebSocket Client to Helidon Websocket Server.

## Impact

Impossible to use JDK WebSocket Client to talk to Helidon WebSocket server. 

## Steps to reproduce

See sample JUnit 5 testcase below:

```java

package io.helidon.webserver.tyrus;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.fail;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.WebSocket;
import java.util.concurrent.BlockingDeque;
import java.util.concurrent.CompletionStage;
import java.util.concurrent.LinkedBlockingDeque;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicReference;
import java.util.logging.Level;
import java.util.logging.Logger;
import javax.websocket.CloseReason;
import javax.websocket.Endpoint;
import javax.websocket.EndpointConfig;
import javax.websocket.MessageHandler;
import javax.websocket.Session;
import javax.websocket.server.ServerEndpointConfig;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

/**
 * Illustrates an issue with premature closing of websocket due to interplay between logic in {@link
 * io.helidon.webserver.ForwardingHandler} and the JDK WebSocket Client adding a spurious {@code
 * Content-Length} header to the websocket upgrade GET request.
 *
 * <p>Note to run this test case the containing {@code module-info.java} MUST have a {@code requires
 * java.net.http;} clause
 *
 */
public class JdkWebSocketClientIssueTest extends TyrusSupportBaseTest {
    private static final Logger LOGGER = Logger.getLogger(EchoEndpointProg.class.getName());
    private static final AtomicReference<Throwable> SERVER_ERROR = new AtomicReference<>();
    private static final BlockingDeque<String> SERVER_RECEIVED_MESSAGES = new LinkedBlockingDeque<>(10);
    private static final BlockingDeque<String> CLIENT_RECEIVED_MESSAGES = new LinkedBlockingDeque<>(10);

    /**
     * JDK WebSocket client message listener
     */
    public static class ClientListener implements WebSocket.Listener {

        public ClientListener() {
        }

        @Override
        public void onOpen(final WebSocket webSocket) {
            webSocket.sendText("Hello World",true);
            WebSocket.Listener.super.onOpen(webSocket);
        }

        @Override
        public CompletionStage<?> onText(final WebSocket webSocket, final CharSequence data,
                final boolean last) {
            /* Record all client messages received */
            CLIENT_RECEIVED_MESSAGES.offer(data.toString());

            return WebSocket.Listener.super.onText(webSocket, data, last);
        }
    }

    /**
     * Tyrus javax.websocket server endpoint listener
     */
    public static class ServerEchoEndpoint extends Endpoint {

        @Override
        public void onOpen(final Session session, final EndpointConfig config) {
            session.addMessageHandler(new MessageHandler.Whole<String>() {
                @Override
                public void onMessage(String message) {
                    try {
                        /* record all messages received */
                        SERVER_RECEIVED_MESSAGES.offer(message);
                        session.getBasicRemote().sendObject(message);       // calls encoder
                    } catch (Exception e) {
                        LOGGER.info(e.getMessage());
                    }
                }
            });
        }

        @Override
        public void onClose(Session session, CloseReason closeReason) {
            /* Expect close to be called AFTER a message has been received. */
            assertFalse(SERVER_RECEIVED_MESSAGES.isEmpty());
            super.onClose(session, closeReason);
        }

        @Override
        public void onError(Session session, Throwable thr) {
            SERVER_ERROR.compareAndSet(null,thr);
        }

    }
    @BeforeAll
    public static void startServer() throws Exception {
        ServerEndpointConfig.Builder builder = ServerEndpointConfig.Builder.create(
                ServerEchoEndpoint.class, "/echo");
        webServer(true, builder.build());
    }

    @Test
    public void testJdkWebSocketClient() {
        try {
            var uri = URI.create("ws://localhost:" + webServer().port() + "/tyrus/echo");
            var listener = new ClientListener();

            HttpClient.newHttpClient().newWebSocketBuilder().buildAsync(uri,listener).get();

            /* Expect to receive a response within 10 seconds */
            var message = CLIENT_RECEIVED_MESSAGES.poll(10, TimeUnit.SECONDS);
            
            if (SERVER_ERROR.get() == null) {
                assertEquals("Hello World", message);
            } else {
                fail(SERVER_ERROR.get());
            }
        } catch (Exception e) {
            fail("Unexpected exception " + e);
        }
    }

}
```

Output for above test-case:

```
/Library/Java/JavaVirtualMachines/jdk-11.0.12.jdk/Contents/Home/bin/java -ea -Djava.util.logging.config.file=/Users/cdivilly/development/github/helidon/webserver/tyrus/target/test-classes/logging-test.properties -Xmx1024m -Dfile.encoding=UTF-8 -Didea.test.cyclic.buffer.size=1048576 -javaagent:/Applications/IntelliJ IDEA CE.app/Contents/lib/idea_rt.jar=56318:/Applications/IntelliJ IDEA CE.app/Contents/bin --patch-module io.helidon.webserver.tyrus=/Users/cdivilly/development/github/helidon/webserver/tyrus/target/test-classes --add-reads io.helidon.webserver.tyrus=ALL-UNNAMED --add-opens io.helidon.webserver.tyrus/io.helidon.webserver.tyrus=ALL-UNNAMED --add-modules io.helidon.webserver.tyrus -classpath /Users/cdivilly/.m2/repository/org/junit/platform/junit-platform-launcher/1.7.0/junit-platform-launcher-1.7.0.jar:/Users/cdivilly/.m2/repository/org/junit/platform/junit-platform-engine/1.7.0/junit-platform-engine-1.7.0.jar:/Users/cdivilly/.m2/repository/org/junit/jupiter/junit-jupiter-engine/5.7.0/junit-jupiter-engine-5.7.0.jar:/Applications/IntelliJ IDEA CE.app/Contents/lib/idea_rt.jar:/Applications/IntelliJ IDEA CE.app/Contents/plugins/junit/lib/junit5-rt.jar:/Applications/IntelliJ IDEA CE.app/Contents/plugins/junit/lib/junit-rt.jar:/Users/cdivilly/development/github/helidon/webserver/tyrus/target/test-classes:/Users/cdivilly/.m2/repository/io/netty/netty-resolver/4.1.73.Final/netty-resolver-4.1.73.Final.jar:/Users/cdivilly/.m2/repository/io/netty/netty-tcnative-classes/2.0.46.Final/netty-tcnative-classes-2.0.46.Final.jar:/Users/cdivilly/.m2/repository/jakarta/xml/bind/jakarta.xml.bind-api/2.3.3/jakarta.xml.bind-api-2.3.3.jar:/Users/cdivilly/.m2/repository/jakarta/activation/jakarta.activation-api/1.2.2/jakarta.activation-api-1.2.2.jar:/Users/cdivilly/.m2/repository/javax/websocket/javax.websocket-api/1.1/javax.websocket-api-1.1.jar:/Users/cdivilly/.m2/repository/org/glassfish/tyrus/tyrus-client/1.17/tyrus-client-1.17.jar:/Users/cdivilly/.m2/repository/org/glassfish/tyrus/tyrus-container-grizzly-client/1.17/tyrus-container-grizzly-client-1.17.jar:/Users/cdivilly/.m2/repository/org/glassfish/grizzly/grizzly-framework/2.3.22/grizzly-framework-2.3.22.jar:/Users/cdivilly/.m2/repository/org/glassfish/grizzly/grizzly-http-server/2.3.22/grizzly-http-server-2.3.22.jar:/Users/cdivilly/.m2/repository/org/glassfish/grizzly/grizzly-http/2.3.22/grizzly-http-2.3.22.jar:/Users/cdivilly/development/github/helidon/webserver/test-support/target/classes:/Users/cdivilly/.m2/repository/org/junit/jupiter/junit-jupiter-api/5.7.0/junit-jupiter-api-5.7.0.jar:/Users/cdivilly/.m2/repository/org/apiguardian/apiguardian-api/1.1.0/apiguardian-api-1.1.0.jar:/Users/cdivilly/.m2/repository/org/opentest4j/opentest4j/1.2.0/opentest4j-1.2.0.jar:/Users/cdivilly/.m2/repository/org/junit/platform/junit-platform-commons/1.7.0/junit-platform-commons-1.7.0.jar:/Users/cdivilly/.m2/repository/org/hamcrest/hamcrest-all/1.3/hamcrest-all-1.3.jar -p /Users/cdivilly/development/github/helidon/webserver/tyrus/target/classes:/Users/cdivilly/development/github/helidon/logging/common/target/classes:/Users/cdivilly/.m2/repository/org/glassfish/tyrus/tyrus-core/1.17/tyrus-core-1.17.jar:/Users/cdivilly/development/github/helidon/config/config/target/classes:/Users/cdivilly/development/github/helidon/tracing/config/target/classes:/Users/cdivilly/.m2/repository/io/netty/netty-codec-http/4.1.73.Final/netty-codec-http-4.1.73.Final.jar:/Users/cdivilly/development/github/helidon/common/reactive/target/classes:/Users/cdivilly/.m2/repository/io/netty/netty-handler/4.1.73.Final/netty-handler-4.1.73.Final.jar:/Users/cdivilly/development/github/helidon/common/service-loader/target/classes:/Users/cdivilly/development/github/helidon/common/mapper/target/classes:/Users/cdivilly/.m2/repository/org/glassfish/tyrus/tyrus-server/1.17/tyrus-server-1.17.jar:/Users/cdivilly/.m2/repository/io/netty/netty-codec/4.1.73.Final/netty-codec-4.1.73.Final.jar:/Users/cdivilly/.m2/repository/org/glassfish/tyrus/tyrus-spi/1.17/tyrus-spi-1.17.jar:/Users/cdivilly/development/github/helidon/webserver/webserver/target/classes:/Users/cdivilly/.m2/repository/io/opentracing/opentracing-noop/0.33.0/opentracing-noop-0.33.0.jar:/Users/cdivilly/.m2/repository/io/netty/netty-transport/4.1.73.Final/netty-transport-4.1.73.Final.jar:/Users/cdivilly/development/github/helidon/media/common/target/classes:/Users/cdivilly/.m2/repository/jakarta/websocket/jakarta.websocket-api/1.1.2/jakarta.websocket-api-1.1.2.jar:/Users/cdivilly/development/github/helidon/common/context/target/classes:/Users/cdivilly/development/github/helidon/common/common/target/classes:/Users/cdivilly/development/github/helidon/common/http/target/classes:/Users/cdivilly/.m2/repository/io/netty/netty-buffer/4.1.73.Final/netty-buffer-4.1.73.Final.jar:/Users/cdivilly/.m2/repository/jakarta/annotation/jakarta.annotation-api/1.3.5/jakarta.annotation-api-1.3.5.jar:/Users/cdivilly/development/github/helidon/common/media-type/target/classes:/Users/cdivilly/.m2/repository/io/opentracing/opentracing-api/0.33.0/opentracing-api-0.33.0.jar:/Users/cdivilly/.m2/repository/io/opentracing/opentracing-util/0.33.0/opentracing-util-0.33.0.jar:/Users/cdivilly/development/github/helidon/common/configurable/target/classes:/Users/cdivilly/development/github/helidon/common/key-util/target/classes:/Users/cdivilly/.m2/repository/io/netty/netty-codec-http2/4.1.73.Final/netty-codec-http2-4.1.73.Final.jar:/Users/cdivilly/.m2/repository/io/netty/netty-common/4.1.73.Final/netty-common-4.1.73.Final.jar com.intellij.rt.junit.JUnitStarter -ideVersion5 -junit5 io.helidon.webserver.tyrus.JdkWebSocketClientIssueTest
2022.03.22 10:26:20 INFO io.helidon.webserver.tyrus.TyrusSupport Thread[main,5,main]: Updating TyrusSupport routing routes
2022.03.22 10:26:21 INFO io.helidon.webserver.NettyWebServer Thread[nioEventLoopGroup-2-1,10,main]: Channel '@default' started: [id: 0xac1827cf, L:/127.0.0.1:56322]

org.opentest4j.AssertionFailedError
	at org.junit.jupiter.api.AssertionUtils.fail(AssertionUtils.java:47)
	at org.junit.jupiter.api.Assertions.fail(Assertions.java:140)
	at io.helidon.webserver.tyrus/io.helidon.webserver.tyrus.JdkWebSocketClientIssueTest.testJdkWebSocketClient(JdkWebSocketClientIssueTest.java:135)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:688)
	at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
	at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:149)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:140)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestMethod(TimeoutExtension.java:84)
	at org.junit.jupiter.engine.execution.ExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(ExecutableInvoker.java:115)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.lambda$invoke$0(ExecutableInvoker.java:105)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:104)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:98)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$6(TestMethodTestDescriptor.java:210)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:206)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:131)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:65)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:129)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:127)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:126)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:84)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:143)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:129)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:127)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:126)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:84)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:143)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:129)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:127)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:126)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:84)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:32)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:51)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:108)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:88)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:67)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:52)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:96)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:75)
	at com.intellij.junit5.JUnit5IdeaTestRunner.startRunnerWithArgs(JUnit5IdeaTestRunner.java:71)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater$1.execute(IdeaTestRunner.java:38)
	at com.intellij.rt.execution.junit.TestsRepeater.repeat(TestsRepeater.java:11)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:35)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:235)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:54)
Caused by: org.opentest4j.AssertionFailedError: expected: <false> but was: <true>
	at org.junit.jupiter.api.AssertionUtils.fail(AssertionUtils.java:55)
	at org.junit.jupiter.api.AssertFalse.assertFalse(AssertFalse.java:40)
	at org.junit.jupiter.api.AssertFalse.assertFalse(AssertFalse.java:35)
	at org.junit.jupiter.api.Assertions.assertFalse(Assertions.java:210)
	at io.helidon.webserver.tyrus/io.helidon.webserver.tyrus.JdkWebSocketClientIssueTest$ServerEchoEndpoint.onClose(JdkWebSocketClientIssueTest.java:104)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at tyrus.core@1.17/org.glassfish.tyrus.core.TyrusEndpointWrapper.onClose(TyrusEndpointWrapper.java:1238)
	at tyrus.core@1.17/org.glassfish.tyrus.core.TyrusWebSocket.onClose(TyrusWebSocket.java:110)
	at tyrus.core@1.17/org.glassfish.tyrus.core.ProtocolHandler.close(ProtocolHandler.java:481)
	at tyrus.core@1.17/org.glassfish.tyrus.core.TyrusWebSocket.close(TyrusWebSocket.java:244)
	at tyrus.core@1.17/org.glassfish.tyrus.core.TyrusWebSocketEngine$TyrusConnection.close(TyrusWebSocketEngine.java:830)
	at io.helidon.webserver.tyrus/io.helidon.webserver.tyrus.TyrusReaderSubscriber.onComplete(TyrusReaderSubscriber.java:122)
	at io.helidon.media.common/io.helidon.media.common.MessageBodyContext$EventingSubscriber.onComplete(MessageBodyContext.java:354)
	at io.helidon.common.reactive@2.4.2-SNAPSHOT/io.helidon.common.reactive.MultiTappedPublisher$MultiTappedSubscriber.onComplete(MultiTappedPublisher.java:252)
	at io.helidon.common.reactive@2.4.2-SNAPSHOT/io.helidon.common.reactive.BufferedEmittingPublisher.drain(BufferedEmittingPublisher.java:465)
	at io.helidon.common.reactive@2.4.2-SNAPSHOT/io.helidon.common.reactive.BufferedEmittingPublisher.subscribe(BufferedEmittingPublisher.java:110)
	at io.helidon.common.reactive@2.4.2-SNAPSHOT/io.helidon.common.reactive.MultiFromPublisher.subscribe(MultiFromPublisher.java:39)
	at io.helidon.common.reactive@2.4.2-SNAPSHOT/io.helidon.common.reactive.MultiTappedPublisher.subscribe(MultiTappedPublisher.java:74)
	at io.helidon.media.common/io.helidon.media.common.MessageBodyContext$EventingPublisher.subscribe(MessageBodyContext.java:293)
	at io.helidon.media.common/io.helidon.media.common.MessageBodyReadableContent.subscribe(MessageBodyReadableContent.java:109)
	at io.helidon.webserver.tyrus/io.helidon.webserver.tyrus.TyrusSupport$TyrusHandler.accept(TyrusSupport.java:315)
	at io.helidon.webserver/io.helidon.webserver.RequestRouting$RoutedRequest.next(RequestRouting.java:324)
	at io.helidon.webserver/io.helidon.webserver.WebTracingConfig$RequestSpanHandler.accept(WebTracingConfig.java:247)
	at io.helidon.webserver/io.helidon.webserver.RequestRouting$RoutedRequest.next(RequestRouting.java:324)
	at io.helidon.webserver/io.helidon.webserver.RequestRouting.route(RequestRouting.java:89)
	at io.helidon.webserver/io.helidon.webserver.ForwardingHandler.lambda$channelReadHttpRequest$18(ForwardingHandler.java:451)
	at io.helidon.webserver/io.helidon.webserver.RequestContext.lambda$runInScope$4(RequestContext.java:72)
	at io.helidon.common.context/io.helidon.common.context.Contexts.runInContext(Contexts.java:117)
	at io.helidon.webserver/io.helidon.webserver.RequestContext.runInScope(RequestContext.java:70)
	at io.helidon.webserver/io.helidon.webserver.ForwardingHandler.channelReadHttpRequest(ForwardingHandler.java:451)
	at io.helidon.webserver/io.helidon.webserver.ForwardingHandler.lambda$channelRead0$3(ForwardingHandler.java:163)
	at io.helidon.common.context/io.helidon.common.context.Contexts.runInContext(Contexts.java:137)
	at io.helidon.webserver/io.helidon.webserver.ForwardingHandler.channelRead0(ForwardingHandler.java:162)
	at io.netty.transport@4.1.73.Final/io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:99)
	at io.netty.transport@4.1.73.Final/io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.transport@4.1.73.Final/io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.transport@4.1.73.Final/io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.codec@4.1.73.Final/io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:327)
	at io.netty.codec@4.1.73.Final/io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:299)
	at io.netty.transport@4.1.73.Final/io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.transport@4.1.73.Final/io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.transport@4.1.73.Final/io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.transport@4.1.73.Final/io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	at io.netty.transport@4.1.73.Final/io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.transport@4.1.73.Final/io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.transport@4.1.73.Final/io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	at io.netty.transport@4.1.73.Final/io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
	at io.netty.transport@4.1.73.Final/io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:722)
	at io.netty.transport@4.1.73.Final/io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:658)
	at io.netty.transport@4.1.73.Final/io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:584)
	at io.netty.transport@4.1.73.Final/io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:496)
	at io.netty.common@4.1.73.Final/io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:986)
	at io.netty.common@4.1.73.Final/io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.common@4.1.73.Final/io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:834)


Process finished with exit code 255
```
- Test case asserts that at least 1 message is received from the client, before the server websocket session is closed. 
- However because of the described issue with the spurious `Content-Length: 0` on the upgrade requests, Helidon marks the request completed, causing the websocket session to be immediately closed, before any messages can be received.

--- a/webserver/tyrus/pom.xml
+++ b/webserver/tyrus/pom.xml
@@ -86,6 +86,24 @@
         <pluginManagement>
             <plugins>
                 <plugin>
+                    <groupId>org.apache.maven.plugins</groupId>
+                    <artifactId>maven-compiler-plugin</artifactId>
+                    <executions>
+                        <execution>
+                            <id>default-testCompile</id>
+                            <configuration>
+                                <!--
+                                We use HTTP client from tests, and we do not want it as a module dependency
+                                 -->
+                                <compilerArgs>
+                                    <compilerArg>--add-modules</compilerArg>
+                                    <compilerArg>java.net.http</compilerArg>
+                                </compilerArgs>
+                            </configuration>
+                        </execution>
+                    </executions>
+                </plugin>
+                <plugin>
                     <artifactId>maven-surefire-plugin</artifactId>
                     <configuration>
                         <systemPropertyVariables>

--- a/webserver/tyrus/src/test/java/io/helidon/webserver/tyrus/HttpClientTest.java
+++ b/webserver/tyrus/src/test/java/io/helidon/webserver/tyrus/HttpClientTest.java
@@ -1,0 +1,85 @@
+/*
+ * Copyright (c) 2022 Oracle and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.helidon.webserver.tyrus;
+
+import java.net.URI;
+import java.net.http.HttpClient;
+import java.net.http.WebSocket;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import javax.websocket.server.ServerEndpointConfig;
+
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.MatcherAssert.assertThat;
+
+class HttpClientTest extends TyrusSupportBaseTest {
+    @BeforeAll
+    static void startServer() throws ExecutionException, InterruptedException, TimeoutException {
+        ServerEndpointConfig.Builder builder = ServerEndpointConfig.Builder.create(
+                EchoEndpoint.class, "/");
+        webServer(true, builder.build());
+    }
+
+    @Test
+    void testJdkClient() throws ExecutionException, InterruptedException, TimeoutException {
+        URI uri = URI.create("ws://localhost:" + webServer().port() + "/tyrus/echo");
+        ClientListener listener = new ClientListener();
+
+        WebSocket webSocket = HttpClient.newHttpClient()
+                .newWebSocketBuilder()
+                .buildAsync(uri, listener)
+                .get();
+
+        webSocket.sendText("message", true);
+        webSocket.sendClose(WebSocket.NORMAL_CLOSURE, "finished");
+
+        String response = listener.await();
+        assertThat(response, is("message"));
+    }
+
+    private static class ClientListener implements WebSocket.Listener {
+        private final CompletableFuture<String> future = new CompletableFuture<>();
+
+        @Override
+        public CompletionStage<?> onText(WebSocket webSocket, CharSequence data, boolean last) {
+            future.complete(data.toString());
+            return CompletableFuture.completedFuture(data);
+        }
+
+        @Override
+        public CompletionStage<?> onClose(WebSocket webSocket, int statusCode, String reason) {
+            future.complete(reason);
+            return CompletableFuture.completedFuture(reason);
+        }
+
+        @Override
+        public void onError(WebSocket webSocket, Throwable error) {
+            future.completeExceptionally(error);
+        }
+
+        public String await() throws ExecutionException, InterruptedException, TimeoutException {
+            return future.get(10, TimeUnit.SECONDS);
+        }
+    }
+}

--- a/webserver/webserver/src/main/java/io/helidon/webserver/ForwardingHandler.java
+++ b/webserver/webserver/src/main/java/io/helidon/webserver/ForwardingHandler.java
@@ -367,7 +367,9 @@ public class ForwardingHandler extends SimpleChannelInboundHandler<Object> {
 
         String contentLength = request.headers().get(HttpHeaderNames.CONTENT_LENGTH);
 
+        // HTTP WebSocket client sends a content length of 0 together with Connection: Upgrade
         if ("0".equals(contentLength)
+                               && !"upgrade".equalsIgnoreCase(request.headers().get(HttpHeaderNames.CONNECTION))
                 || (contentLength == null
                              && !"upgrade".equalsIgnoreCase(request.headers().get(HttpHeaderNames.CONNECTION))
                              && !"chunked".equalsIgnoreCase(request.headers().get(HttpHeaderNames.TRANSFER_ENCODING))

